clients/da-vinci-client/src/main/java/com/linkedin/davinci/blobtransfer/NettyP2PBlobTransferManager.java

```diff
@@ -26,147 +25,179 @@
-            retryCount++;
-          } else if (e.getCause() instanceof VeniceBlobTransferFileNotFoundException) {
-            // error case 3: the connected host does not have the requested file,
-            // solution: move to next possible host
-            LOGGER.warn(
-                "Peer {} does not have the requested blob for store {} version {} partition {}, moving to next possible host.",
-                peer,
-                storeName,
-                version,
-                partition,
-                e);
-            break;
-          } else {
-            // error case 4:
-            // other exceptions (InterruptedException, ExecutionException, TimeoutException) that are not expected,
-            // solution: do not use blob transfer to bootstrap at all for saving time
-            String errorMessage = String.format(
-                "Failed to connect to peer %s for partition %d store %s version %d with exception. "
-                    + "Skip bootstrap the partition from blob transfer.",
-                peer,
-                partition,
-                storeName,
-                version);
-            LOGGER.error(errorMessage, e);
-            throw new VeniceException(errorMessage, e);
-          }
+    // Iterate through each peer and chain the futures
+    for (int currentPeerIndex = 0; currentPeerIndex < peers.size(); currentPeerIndex++) {
+      final int peerIndex = currentPeerIndex;
+      // Chain the next operation to the previous future
+      chainOfPeersFuture = chainOfPeersFuture.thenCompose(v -> {
+        String chosenHost = peers.get(peerIndex).split("_")[0];
+
+        if (resultFuture.isDone()) {
+          LOGGER.info(String.format(SKIP_FETCHED_BLOB_MSG_FORMAT, chosenHost, storeName, partition, version));
+          return CompletableFuture.completedFuture(null);
         }
-      }
 
-      LOGGER.warn(
-          "Failed to connect to peer {} for partition {} store {} version {} after {} attempts, "
-              + "moving to next possible host to bootstrap the partition.",
-          peer,
-          partition,
-          storeName,
-          version,
-          retryCount);
+        // Attempt to fetch the blob from the current peer asynchronously
+        LOGGER.info("Attempting to connect to host: {}", chosenHost);
+
+        return nettyClient.get(chosenHost, storeName, version, partition)
+            .toCompletableFuture()
+            .thenAccept(inputStream -> {
+              // Success case: Complete the future with the input stream
+              LOGGER.info(
+                  String.format(
+                      FETCHED_BLOB_SUCCESS_MSG_FORMAT,
+                      chosenHost,
+                      storeName,
+                      partition,
+                      version,
+                      Duration.between(startTime, Instant.now()).getSeconds()));
+              resultFuture.complete(inputStream);
+            })
+            .exceptionally(ex -> {
```

Comment:
I am a bit lost in the future chaining, could you clarify about this: thenAccept returns a new future, but exceptionally  is applied on the new future, instead of the original future. Is my understanding correct?

Reply:
`thenAccept` is called when the CompletableFuture completes successfully (i.e., without throwing an exception), whereas `exceptionally` is invoked when it completes with an exception.

These represent two exclusive cases: one for success and one for exceptions. The `exceptionally` method is used to handle exceptions from the original future, not from the new future created by `thenAccept`.

Reply:
Synced offline, you are correct. The thenApply won't be executed and no side effect is generated so exceptioanlly should apply on original completion stage or future.